### PR TITLE
Add Type checking and enforce lint

### PR DESCRIPTION
**Pull request by vdeturckheim**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -25,12 +25,13 @@ jobs:
...
         pip install -r requirements.txt
+    - name: Type check with mypy
+      run: mypy --install-types --non-interactive guarddog
     - name: Lint with flake8
       run: |
         # stop the build if there are Python syntax errors or undefined names
-        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics --exclude tests/analyzer/sourcecode,tests/analyzer/metadata/resources,evaluator/data
-        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
-        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics --exclude tests/analyzer/sourcecode,tests/analyzer/metadata/resources,evaluator/data --ignore=E203,W503
+        flake8 guarddog --count --select=E9,F63,F7,F82 --show-source --statistics --exclude tests/analyzer/sourcecode,tests/analyzer/metadata/resources,evaluator/data
```
Comment (christophetd) [inline .github/workflows/test.yml:33]: same here

**Review by christophetd — COMMENTED** [review]
Looks great! Can we add a pre-commit hook configuration through https://pre-commit.com/ to allow for automatic linting and typing check?

Also do we need any kind of IDE setup docs? (I don't think so if using PyCharm?)

File: .github/workflows/test.yml (hunk 1 above)
Comment (vdeturckheim) [reply .github/workflows/test.yml]: done

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -25,12 +25,13 @@ jobs:
       run: |
         python -m pip install --upgrade pip
         pip install -r requirements.txt
+    - name: Type check with mypy
```
[the remark this answers is not in the document]
Comment (vdeturckheim) [reply .github/workflows/test.yml]: done

**Comment by vdeturckheim** [comment]
Adding do now.

Regarding pre-commit hooks. I am not a big fan, more than once, I wanted to commit something as a state save and hooks went in my way.

**Comment by christophetd** [comment]
You don't need to install them - you can even run them manually - but it helps running the same checks at the CI pipeline. Can we at least put it in a Makefile or something so people can run `make test` manually locally and be confident the CI will pass?

**Comment by vdeturckheim** [comment]
Done!

**Review by christophetd — APPROVED** [review]

**Pull request merged**